dask/dataframe/io/parquet/utils.py

```diff
@@ -10,6 +13,86 @@
 class Engine:
     """The API necessary to provide a new Parquet reader/writer"""
 
+    @classmethod
+    def extract_filesystem(
+        cls,
+        urlpath,
+        filesystem,
+        dataset_options,
+        open_file_options,
+        storage_options,
+    ):
+        """Extract filesystem object from urlpath or user arguments
+
+        This classmethod should only be overridden for engines that need
+        to handle filesystem implementations other than ``fsspec``
+        (e.g. ``pyarrow.fs.S3FileSystem``).
+
+        Parameters
+        ----------
+        urlpath: str or List[str]
+            Source directory for data, or path(s) to individual parquet files.
+        filesystem: "fsspec" or fsspec.AbstractFileSystem
+            Filesystem backend to use. Default is "fsspec"
+        dataset_options: dict
+            Engine-specific dataset options.
+        open_file_options: dict
+            Options to be used for file-opening at read time.
+        storage_options: dict
+            Options to be passed on to the file-system backend.
+
+        Returns
+        -------
+        fs: Any
+            A global filesystem object to be used for metadata
+            processing and file-opening by the engine.
+        paths: List[str]
+            List of data-source paths.
+        dataset_options: dict
+            Engine-specific dataset options.
+        open_file_options: dict
+            Options to be used for file-opening at read time.
+        """
+
+        # Check if fs was specified as a dataset option
+        fs = dataset_options.pop("fs", "fsspec")
+        if filesystem is not None:
+            fs = filesystem
+
+        if fs in (None, "fsspec"):
+            # Use fsspec to infer a filesystem by default
+            fs, _, paths = get_fs_token_paths(
+                urlpath, mode="rb", storage_options=storage_options
+            )
+            return fs, paths, dataset_options, open_file_options
+
+        else:
+            # Check that an initialized filesystem object was provided
+            if not isinstance(fs, AbstractFileSystem):
+                raise ValueError(
+                    f"Expected fsspec.AbstractFileSystem or 'fsspec'. Got {fs}"
+                )
+
+            if storage_options:
+                # The filesystem was already specified. Can't pass in
+                # any storage options
+                warnings.warn(f"Ignoring storage_options: {storage_options}")
```

Comment:
I'd prefer to make this an error and fail loudly instead of continuing on as that seems less likely to lead to confusion

Reply:
sounds good